### PR TITLE
Add Invoke-History support for invoking a collection of commands

### LINKED ISSUE #12488: Invoke-History should be able to invoke more than one command
# Summary of the new feature/enhancement

My exact use case is:
I have a series of commands in my history which I want to run one after another for example
```
stop-x
copy-xlog
delete-xlog
start-x
```
These might be 101,102,103,104 in the history 

one might expect to be able to run either 
`101..104 | invoke-history`
or 
Ihy 101,102,103,104

The former causes an error
`Invoke-History: The input object cannot be bound to any parameters for the command either because the command does not take pipeline input `
and then invokes the previous command

The latter causes an error `Invoke-History: Cannot convert 'System.Object[]' to the type 'System.String'`

Ideally any cmdlet should support its "target" * coming from the pipeline and support a comma-separated list of targets.  A few which could, don't, without an obvious reason to be an exception.  This is a proposal to fix one of those.
 
\* Target being the parameter telling the command what to work **on**, where other parameters  the command **how** the work should be done

### PR DESCRIPTION
**Pull request by thomasrayner**
# PR Summary

Resolves #12488.

Adds support for invoking a collection of command IDs while using `Invoke-History`.

## PR Context

When there are a series of commands in history, it is often useful to execute several of them at once in series. Currently, the `-Id` parameter on `Invoke-History` is a `string`. This PR changes it to a `string[]` and adds a loop to process a collection of IDs instead of a single ID. This also adds some consistency with `Get-History` whose `-Id` parameter does support collections of IDs.

Previously:
Previously

After this PR:
After this PR:

There are a variety of other ways to implement this kind of change, but I thought it was best to perform the same checking and validation per ID so that if some IDs were valid but others weren't, the valid IDs up until the invalid ID still run.

Example:
Example

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: MicrosoftDocs/PowerShell-Docs/issues/6250
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/hostifaces/History.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/hostifaces/History.cs:1018]: This will throw with Invoke-History without arguments.

Comment (thomasrayner) [reply src/System.Management.Automation/engine/hostifaces/History.cs]: Would you like me to add a test for calling `Invoke-History` without any args? At the moment, I don't think there is one.

Comment (iSazonov) [reply src/System.Management.Automation/engine/hostifaces/History.cs]: Yes, we need to cover the case by a test.

File: src/System.Management.Automation/engine/hostifaces/History.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/hostifaces/History.cs:981]: Please remove the commented code.

File: src/System.Management.Automation/engine/hostifaces/History.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/hostifaces/History.cs:1027]: Please remove the resource string from Resx file.

File: src/System.Management.Automation/engine/hostifaces/History.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/hostifaces/History.cs:1010]: What if History empty?

File: src/System.Management.Automation/engine/hostifaces/History.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/hostifaces/History.cs:1053]:
```suggestion
                if (!ShouldProcess(commandToInvoke))
```

File: src/System.Management.Automation/engine/hostifaces/History.cs
Comment (thomasrayner) [reply src/System.Management.Automation/engine/hostifaces/History.cs:1010]: Are we protected against that by L999 `Dbg.Assert(history != null, "History should be non null");`? Or is that inadequate here?

Comment (vexx32) [reply src/System.Management.Automation/engine/hostifaces/History.cs:1005]: That will only protect debug builds. If this is an expected result in release builds we'll need to write a more normal error instead (and probably just remove the assertion).

Or it could simply "do nothing" if there is no history, but even there you'll probably need to check for it and return early in that case.

Comment (iSazonov) [reply src/System.Management.Automation/engine/hostifaces/History.cs:1005]: Simple test - run pwsh and run first command `Invoke-History`.

Comment (thomasrayner) [reply src/System.Management.Automation/engine/hostifaces/History.cs:1005]: image

Yeah... that should probably be a better error message. I'll change that.

Comment (iSazonov) [reply src/System.Management.Automation/engine/hostifaces/History.cs:1005]: See a message before the change.

Comment (thomasrayner) [reply src/System.Management.Automation/engine/hostifaces/History.cs]: Sorry, @iSazonov, I'm not sure what exactly you mean there.

But, I've changed how the error is thrown and added a test to validate the type of error that gets thrown when there's no history.

**Comment by daxian-dbw** [comment]
Remove the `review-needed` label as the PR is WIP.

**Comment by thomasrayner** [comment]
Is there any recommendation on what kind of doc update is required for this PR? Should there be a new help example, something in docs.microsoft.com, etc? Or is it all mostly auto-generated?

**Comment by vexx32** [comment]
I think some of it will be autogenerated but at the least an issue will need to be filed in the docs repo to ensure that the parameter is added with a description etc., alongside any necessary examples.

**Comment by thomasrayner** [comment]
> I think some of it will be autogenerated but at the least an issue will need to be filed in the [docs repo](/MicrosoftDocs/PowerShell-Docs) to ensure that the parameter is added with a description etc., alongside any necessary examples.

Does that issue with the docs repo need to be resolved before this can merge? I'm just trying to figure out what I *have* to do in order to finish this PR.

**Comment by vexx32** [comment]
Nope, as long as there's a tracking issue for any required docs the code is generally OK to be merged in.

**Comment by thomasrayner** [comment]
I messed it up trying to satisfy CodeFactor. What's the right way to organize the code it's complaining about? (L1093 - L1115)

File: src/System.Management.Automation/engine/hostifaces/History.cs
Comment (vexx32) [inline src/System.Management.Automation/engine/hostifaces/History.cs:1104]:
```suggestion
                    EventHandler<DataAddedEventArgs> debugAdded = delegate (object sender, DataAddedEventArgs e)
                    {
                        DebugRecord record = (DebugRecord)((PSDataCollection<DebugRecord>)sender)[e.Index]; WriteDebug(record.Message);
                    };
                    EventHandler<DataAddedEventArgs> errorAdded = delegate (object sender, DataAddedEventArgs e)
                    {
                        ErrorRecord record = (ErrorRecord)((PSDataCollection<ErrorRecord>)sender)[e.Index]; WriteError(record);
                    };
                    EventHandler<DataAddedEventArgs> informationAdded = delegate (object sender, DataAddedEventArgs e)
                    {
                        InformationRecord record = (InformationRecord)((PSDataCollection<InformationRecord>)sender)[e.Index]; WriteInformation(record);
                    };
                    EventHandler<DataAddedEventArgs> progressAdded = delegate (object sender, DataAddedEventArgs e)
                    {
                        ProgressRecord record = (ProgressRecord)((PSDataCollection<ProgressRecord>)sender)[e.Index]; WriteProgress(record);
                    };
                    EventHandler<DataAddedEventArgs> verboseAdded = delegate (object sender, DataAddedEventArgs e)
                    {
                        VerboseRecord record = (VerboseRecord)((PSDataCollection<VerboseRecord>)sender)[e.Index]; WriteVerbose(record.Message);
                    };
                    EventHandler<DataAddedEventArgs> warningAdded = delegate (object sender, DataAddedEventArgs e)
                    {
                        WarningRecord record = (WarningRecord)((PSDataCollection<WarningRecord>)sender)[e.Index]; WriteWarning(record.Message);
                    };
```

@thomasrayner  I _think_ this is what it's looking for?

Comment (thomasrayner) [reply src/System.Management.Automation/engine/hostifaces/History.cs:1104]: @vexx32 looks like what I had in this commit which didn't build: https://github.com/PowerShell/PowerShell/commit/711a90e2ce90fc5caefa5c48856f3c640a038129

Comment (vexx32) [reply src/System.Management.Automation/engine/hostifaces/History.cs:1104]: @thomasrayner looks like you accidentally cut out one of the `delegate` lines in your version, which is why it didn't build. I've noticed mine could be a little better, I'll edit the suggestion.

Comment (thomasrayner) [reply src/System.Management.Automation/engine/hostifaces/History.cs:1104]: @vexx32 WELL THAT WOULD DO IT :joy:

**Comment by vexx32** [comment]
@PoshChan please retry macos

**Comment by PoshChan** [comment]
@vexx32, did not find matching build context: `PowerShell-CI-macOS`; allowed contexts: PowerShell-CI-SSH

**Comment by daxian-dbw** [comment]
@iSazonov and @vexx32 can you please continue with the review when you get time? Thanks!

File: src/System.Management.Automation/engine/hostifaces/History.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/hostifaces/History.cs:989]: Should we have an issue to update SDK?

File: src/System.Management.Automation/engine/hostifaces/History.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/hostifaces/History.cs:1025]: nit: named argument.

File: src/System.Management.Automation/engine/hostifaces/History.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/hostifaces/History.cs:1041]:
```suggestion
                if (!pipeline.PresentInInvokeHistoryEntryList(entry))
```

File: src/System.Management.Automation/engine/hostifaces/History.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/hostifaces/History.cs:1060]: nit: named argument.

**Review by iSazonov — APPROVED** [review]
With some minor comments.

File: src/System.Management.Automation/engine/hostifaces/History.cs
Comment (thomasrayner) [reply src/System.Management.Automation/engine/hostifaces/History.cs:989]: @iSazonov, not sure what the process is for that.

**Comment by thomasrayner** [comment]
Negotiating this merge conflict is going to take some doing.

File: src/System.Management.Automation/engine/hostifaces/History.cs
Comment (vexx32) [inline src/System.Management.Automation/engine/hostifaces/History.cs:1032]: nit: named arguments for `history.GetEntries()`

File: test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1
Comment (vexx32) [inline test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1:43]: Am I misremembering or does Invoke-History also accept piped IDs? Probably best to have that included in at least once test if so.

**Review by vexx32 — APPROVED** [review]
Looks good to me! Only a few minor comments, nothing blocking.

File: src/System.Management.Automation/engine/hostifaces/History.cs
Comment (thomasrayner) [reply src/System.Management.Automation/engine/hostifaces/History.cs:1032]: Hm, I see your point. Is this a style thing up for debate or not? I get why it's worthwhile to do for non-obvious things like `null` but I wonder if it really adds any notable value in a situation like this?

File: test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1
Comment (thomasrayner) [reply test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1:43]: I don't think it does. This is the kind of experience you get.

```powershell
PS /workspaces/PowerShell> clear-history
PS /workspaces/PowerShell> write-output "hi"        
hi
PS /workspaces/PowerShell> write-output "lo"        
lo
PS /workspaces/PowerShell> get-history

  Id     Duration CommandLine
  --     -------- -----------
  21        0.001 clear-history
  22        0.001 write-output "hi"
  23        0.001 write-output "lo"

PS /workspaces/PowerShell> @(22,23) | invoke-history
Invoke-History: The input object cannot be bound to any parameters for the command either because the command does not take pipeline input or the input and its properties do not match any of the parameters that take pipeline input.
Invoke-History: The input object cannot be bound to any parameters for the command either because the command does not take pipeline input or the input and its properties do not match any of the parameters that take pipeline input.
get-history

  Id     Duration CommandLine
  --     -------- -----------
  21        0.001 clear-history
  22        0.001 write-output "hi"
  23        0.001 write-output "lo"
  24        0.003 get-history

PS /workspaces/PowerShell> 
```

File: src/System.Management.Automation/engine/hostifaces/History.cs
Comment (vexx32) [reply src/System.Management.Automation/engine/hostifaces/History.cs:1032]: 🤷 I tend to prefer it myself, but defer to others here whether it's strictly necessary.

At a glance, though, I can only infer two out of three of those arguments from context.

File: src/System.Management.Automation/engine/hostifaces/History.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/hostifaces/History.cs:1013]: The quest is more for MSFT team. /cc @daxian-dbw

File: src/System.Management.Automation/engine/hostifaces/History.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/hostifaces/History.cs:1032]: @thomasrayner It is minor comment and It's up to you.

**Comment by iSazonov** [comment]
@daxian-dbw The PR is ready for your review and merge.

File: src/System.Management.Automation/engine/hostifaces/History.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/hostifaces/History.cs:1041]: Instead of doing this, the logic should be built in `GetHistoryEntryToInvoke`, so that `GetHistoryEntryToInvoke` will return an `IEnumerable<HistoryInfo>` instead of a single `HistoryInfo`.

File: src/System.Management.Automation/engine/hostifaces/History.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/hostifaces/History.cs:1085]: Today `Invoke-Command` will replace the current history with the one it's about invoke. That behavior makes sense when it only supports the invocation of one history entry.

After the change, `Invoke-Command` can invoke multiple command entries, and thus the current history will be replaced with the last command entry. The behavior doesn't make sense IMO when invoke multiple histories. Maybe in case of invoking multiple history entries, we shouldn't change the history string.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/hostifaces/History.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/hostifaces/History.cs:1044]: I know this is the style used a lot in this file, but we prefer to the below style in this code base.
So appreciate if you can use the below style :)
```
                    ThrowTerminatingError(
                        new ErrorRecord(
                            new InvalidOperationException(HistoryStrings.InvokeHistoryLoopDetected),
                            "InvokeHistoryLoopDetected",
                            ErrorCategory.InvalidOperation,
                            targetObject: null));
```

File: src/System.Management.Automation/engine/hostifaces/History.cs
Comment (vexx32) [reply src/System.Management.Automation/engine/hostifaces/History.cs:1085]: Yeah, I reckon it should just leave the Invoke-History line itself in the history, there isn't really a better way to capture it (unless you wanted to concatenate _all_ the invoked history entries and combine them into one history entry, but I don't think that's really necessary)

**Comment by thomasrayner** [comment]
> This pull request has been automatically marked as stale because it has been marked as requiring author feedback but has not had any activity for **15 days**. It will be closed if no further activity occurs **within 10 days of this comment**.

Heh, I'm still here! It's just been tough to prioritize this with so much on my plate at the moment. Sorry for the delay.

**Comment by thomasrayner** [comment]
I messed up my local git history, and now my fork's history. I'll re-open this later with a clean history. Apologies.

**Pull request closed without merging**
